src/snowflake/snowpark/_internal/data_source_utils.py

```diff
@@ -283,7 +286,10 @@ def infer_data_source_schema(
 
 def data_source_data_to_pandas_df(data: List[Any], schema: StructType) -> pd.DataFrame:
     columns = [col.name for col in schema.fields]
-    df = pd.DataFrame.from_records(data, columns=columns)
+    if not data:
+        return pd.DataFrame(columns=columns)
+    # this way handles both list of object and list of tuples and avoid implict pandas type conversion
+    df = pd.DataFrame([list(row) for row in data], columns=columns, dtype=object)
```

Comment:
I think we also need to test if there is any perf degrade after we do this

Reply:
good call out, doing it now

Reply:
test with no perf degrade